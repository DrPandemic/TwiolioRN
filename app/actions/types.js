// @flow

import { PhoneNumber, Message } from '../types';

export const SET_FETCHED_ACCOUNT_NUMBERS = 'SET_FETCHED_ACCOUNT_NUMBERS';
export const FETCH_ACCOUNT_NUMBERS = 'FETCH_ACCOUNT_NUMBERS';
export const FETCH_ACCOUNT_NUMBER_ERROR = 'FETCH_ACCOUNT_NUMBER_ERROR';

export const SELECT_NUMBER = 'SELECT_NUMBER';
export const SELECT_CONVERSATION = 'SELECT_CONVERSATION';

export const FETCH_MESSAGES = 'FETCH_MESSAGES';
export const SET_FETCHED_MESSAGES = 'SET_FETCHED_MESSAGES';
export const FETCH_MESSAGE_ERROR = 'FETCH_MESSAGE_ERROR';

export type FetchAccountNumbersT = {|
  type: 'FETCH_ACCOUNT_NUMBERS',
|}
export type SuccessFetchAccountNumbersT = {|
  type: 'SET_FETCHED_ACCOUNT_NUMBERS',
  fetchedAccountNumbers: Array<PhoneNumber>,
|}
export type FailFetchAccountNumbersT = {|
    type: 'FETCH_ACCOUNT_NUMBER_ERROR',
    error: any,
|}

export type SelectNumberT = {|
  type: 'SELECT_NUMBER',
  selectedNumber: ?string,
|}

export type FetchMessagesT = {|
  type: 'FETCH_MESSAGES',
|}
export type SuccessFetchMessagesT = {|
    type: 'SET_FETCHED_MESSAGES',
    fetchedMessages: Array<Message>,
|}
export type FailFetchMessagesT = {|
    type: 'FETCH_MESSAGE_ERROR',
    error: any,
|}

export type ActionT =
  | FetchAccountNumbersT
  | SuccessFetchAccountNumbersT
  | FailFetchAccountNumbersT
  | SelectNumberT
  | SelectConversationT
  | FetchMessagesT
  | SuccessFetchMessagesT
  | FailFetchMessagesT;
