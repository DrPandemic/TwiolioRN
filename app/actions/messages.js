// @flow

import * as types from './types';
import { Message } from '../types';

export function fetchMessages(): types.FetchMessagesT {
  return {
    type: types.FETCH_MESSAGES,
  };
}

export function successFetchMessages(messages: Array<Message>):
types.SuccessFetchMessagesT {
  return {
    type: types.SET_FETCHED_MESSAGES,
    fetchedMessages: messages,
  };
}

export function failFetchMessages(error: any):
types.FailFetchMessagesT {
  return {
    type: types.FETCH_MESSAGE_ERROR,
    error,
  };
}
