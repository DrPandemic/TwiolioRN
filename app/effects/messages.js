// @flow

import { Message } from '../types';
import { FetchMessageThresholdInMinutes } from '../constants';

function fetchNextPage(
  api: any,
  url: ?string,
  expandRoute: boolean,
  messages: Array<Message>
): Promise<Array<Message>> {
  if (url === null) {
    return Promise.resolve(messages);
  }

  // Note: If one page fails, all the data is "lost"
  return api.get(url, null, null, expandRoute)
    .then(r => r.json())
    .then(r => fetchNextPage(api, r.next_page_uri, false,
      messages.concat(r.messages.map(message => new Message(message)))));
}

function pad(val: number) {
  return (`00${val}`).slice(-2);
}

function generateDateSent(lastFetch: ?Date): string {
  if (lastFetch != null) {
    let copiedLastFetch = new Date(lastFetch.getTime());
    copiedLastFetch.setMonth(copiedLastFetch.getMonth() - 1);
    copiedLastFetch.setMinutes(
      copiedLastFetch.getMinutes() - FetchMessageThresholdInMinutes
    );

    const month = pad(copiedLastFetch.getMonth() + 1);
    const day = pad(copiedLastFetch.getDate());

    return encodeURI(
      `?DateSent>=${copiedLastFetch.getFullYear()}-${month}-${day}`
    );
  }

  return '';
}

export function fetchMessages(
  api: any,
  lastFetch: ?Date = null,
): Promise<Array<Message>> {
  return fetchNextPage(
    api, `/Messages.json${generateDateSent(lastFetch)}`, true, []
  );
}

export function sendMessage(
  api: any,
  to: string,
  from: string,
  body: string,
): Promise<Message> {
  return api.post('/Messages.json', { To: to, From: from, Body: body })
    .then(r => r.json())
    .then(r => new Message(r));
}
