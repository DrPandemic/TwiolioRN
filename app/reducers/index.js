import { combineReducers } from 'redux';
import * as AccountReducers from './account'

export default combineReducers(Object.assign({},
  AccountReducers,
));
