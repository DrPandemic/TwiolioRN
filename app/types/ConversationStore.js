// @flow

import { Message } from './';
import type { ConversationUsers } from './Message';
import match from '../lib/match';

export type ConversationStoreT = {
  [conversationId: string]: Array<Message>
};

export function addMessage(store: ConversationStoreT, message: Message):
ConversationStoreT {
  const id = message.conversationId;

  return {
    ...store,
    [id]: [message, ...(store[id] || []).filter(m => m.sid !== message.sid)]
  };
}

export function addMessages(
  store: ConversationStoreT,
  messages: Array<Message>
): ConversationStoreT {
  return messages.reduce((s, m) => addMessage(s, m), store);
}

export function getMessages(
  store: ConversationStoreT,
  conversationUsers: ConversationUsers
): Array<Message> {
  return store[Message.getConversationId(conversationUsers)];
}
