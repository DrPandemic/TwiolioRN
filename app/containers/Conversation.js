// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {
  View,
  Text,
} from 'react-native';

import { ActionCreators } from '../actions';
import { PhoneNumber } from '../actions/types';

class AppContainer extends Component {
  props: {};

  render() {
    return (
      <View>
        <Text>
          something
        </Text>
      </View>
    );
  }
}

function mapStateToProps(state) {
  return {};
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(ActionCreators, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(AppContainer);
