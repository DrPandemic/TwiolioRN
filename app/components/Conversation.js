// @flow

import React, { Component } from 'react';
import {
  Dimensions,
  FlatList,
  StyleSheet,
  Text,
  View,
} from 'react-native';
import { List, ListItem } from 'react-native-elements';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { ActionCreators } from '../actions';
import { Message } from '../types';
import { getMessagesById } from '../types/ConversationStore';
import { Colors } from '../constants';
import type { StateT } from '../reducers';
import type { T as MessagesT } from '../reducers/messages';

const screenHeight = Dimensions.get(`window`).height;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    backgroundColor: Colors.background,
    height: screenHeight,
  },
  list: {
    marginTop: 0,
  },
  flatList: {
    marginBottom: 30,
  },
  ourMessage: {
    margin: 10,
    marginLeft: 50,
    marginRight: 30,
  },
  otherMessage: {
    margin: 10,
    marginRight: 50,
  },
  ourTitle: {
    textAlign: 'right',
  },
  otherTitle: {
    color: Colors.primary.text,
  },
  ourMessageBody: {
    backgroundColor: '#fff',
    alignSelf: 'flex-end',
    borderRadius: 10,
    padding: 10,
  },
  otherMessageBody: {
    backgroundColor: Colors.primary.light,
    alignSelf: 'flex-start',
    borderRadius: 10,
    padding: 10,
  },
});

type PropsT = {
  messages: MessagesT,
  location: { state: any }
}

export class PConversation extends Component {
  props: PropsT;

  static renderUs(message: Message) {
    return (
      <ListItem
        key={message.sid}
        title={
          <View style={styles.ourMessageBody}>
            <Text style={styles.ourTitle}>
              {message.body}
            </Text>
          </View>
        }
        style={styles.ourMessage}
        hideChevron={true}
      />
    );
  }

  static renderOther(message: Message) {
    return (
      <ListItem
        key={message.sid}
        title={
          <View style={styles.otherMessageBody}>
            <Text style={styles.otherTitle}>
              {message.body}
            </Text>
          </View>
        }
        style={styles.otherMessage}
        hideChevron={true}
      />
    );
  }

  static renderRow(message: Message) {
    if (message.isInbound) {
      return PConversation.renderUs(message);
    }
    return PConversation.renderOther(message);
  }

  render() {
    const store = this.props.messages.messages;
    const conversationId = this.props.location.state;

    return (
      <View style={styles.container}>
        <List style={styles.list}>
          <FlatList
            data={getMessagesById(store, conversationId)}
            renderItem={({item}) => PConversation.renderRow(item)}
            keyExtractor={message => message.sid}
            style={styles.flatList}
          />
        </List>
      </View>
    );
  }
}

function mapStateToProps(state: StateT) {
  return {
    messages: state.messages,
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(ActionCreators, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(PConversation);
