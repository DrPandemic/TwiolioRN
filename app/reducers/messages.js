// @flow

import { loop, Effects } from 'redux-loop';

import * as types from '../actions/types';
import type { ConversationStoreT } from '../types/ConversationStore';
import { addMessages } from '../types/ConversationStore';
import LibApi from '../lib/api';
import createReducer from '../lib/createReducer';
import effects from '../effects';

export type T = {|
  loading: boolean,
  messages: ConversationStoreT,
  error: any,
|};
export const initialState: T = {
  loading: false,
  messages: {},
  error: {},
};

export const reducer = createReducer({
  [types.FETCH_MESSAGES](state: T) {
    return loop(
      { ...state, loading: true },
      Effects.promise(effects.fetchMessages, LibApi)
    );
  },
  [types.SET_FETCHED_MESSAGES](
    state: T,
    action: types.SuccessFetchMessagesT
  ) {
    return {
      ...state,
      messages: addMessages(state.messages, action.fetchedMessages),
      error: {},
      loading: false,
    };
  },
  [types.FETCH_MESSAGE_ERROR](
    state: T,
    action: types.FailFetchMessagesT
  ) {
    return {
      ...state,
      error: action.error,
      loading: false,
    };
  },
});
