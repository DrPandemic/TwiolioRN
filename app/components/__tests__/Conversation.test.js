// @flow

import React from 'react';
import renderer from 'react-test-renderer';

import { PConversation } from '../Conversation';
import messageFixture from '../../test_helpers/fixtures/received_message.json';
import { Message } from '../../types';
import { addMessages, getMessagesById } from '../../types/ConversationStore';
import { initialState as mInitial } from '../../reducers/messages';

test('renders a message', () => {
  const message = new Message(messageFixture.simple);

  expect(renderer.create(PConversation.renderOther(message))).toMatchSnapshot();
  expect(renderer.create(PConversation.renderUs(message))).toMatchSnapshot();
});

test('renders an empty conversation', () => {
  const list = renderer.create(
      <PConversation
        messages={{ ...mInitial }}
        location={{ state: 'foo' }}
      >
      </PConversation>
  ).toJSON();

  expect(list).toMatchSnapshot();
});

test('renders an non-empty conversation', () => {
  const message = new Message(messageFixture.simple);
  const store = addMessages({}, [
    message,
    new Message(messageFixture.simpleInverse),
    new Message(messageFixture.simpleOtherFrom),
    new Message(messageFixture.secondOutbound),
  ]);

  const list = renderer.create(
      <PConversation
        messages={{
          ...mInitial,
          messages: store,
        }}
        location={{ state: message.conversationId }}
      >
      </PConversation>
  ).toJSON();

  expect(list).toMatchSnapshot();
});
