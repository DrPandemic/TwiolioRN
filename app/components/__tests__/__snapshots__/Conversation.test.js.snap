// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a message 1`] = `
<View
  onLongPress={undefined}
  onPress={undefined}
  style={
    Object {
      "margin": 10,
      "marginRight": 50,
    }
  }
  underlayColor="white"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginLeft": 10,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={undefined}
      >
        <View>
          <View
            style={
              Object {
                "alignSelf": "flex-start",
                "backgroundColor": "#d05ce3",
                "borderRadius": 10,
                "padding": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              disabled={false}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#ffffff",
                }
              }
            >
              foo
            </Text>
          </View>
        </View>
      </View>
      <View
        style={undefined}
      >
        <View />
      </View>
    </View>
  </View>
</View>
`;

exports[`renders a message 2`] = `
<View
  onLongPress={undefined}
  onPress={undefined}
  style={
    Object {
      "margin": 10,
      "marginLeft": 50,
      "marginRight": 30,
    }
  }
  underlayColor="white"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "marginLeft": 10,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={undefined}
      >
        <View>
          <View
            style={
              Object {
                "alignSelf": "flex-end",
                "backgroundColor": "#fff",
                "borderRadius": 10,
                "padding": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              disabled={false}
              ellipsizeMode="tail"
              style={
                Object {
                  "textAlign": "right",
                }
              }
            >
              foo
            </Text>
          </View>
        </View>
      </View>
      <View
        style={undefined}
      >
        <View />
      </View>
    </View>
  </View>
</View>
`;

exports[`renders an empty conversation 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#eee",
      "flexDirection": "column",
      "height": 1334,
    }
  }
>
  <View
    style={
      Object {
        "marginTop": 0,
      }
    }
  >
    <RCTScrollView
      data={Array []}
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      onViewableItemsChanged={undefined}
      renderItem={[Function]}
      renderScrollComponent={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "marginBottom": 30,
        }
      }
      updateCellsBatchingPeriod={50}
      windowSize={21}
    >
      <View />
    </RCTScrollView>
  </View>
</View>
`;

exports[`renders an non-empty conversation 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#eee",
      "flexDirection": "column",
      "height": 1334,
    }
  }
>
  <View
    style={
      Object {
        "marginTop": 0,
      }
    }
  >
    <RCTScrollView
      data={
        Array [
          Message {
            "body": "bar",
            "dateCreated": 2017-05-12T02:46:30.000Z,
            "dateSent": 2017-05-12T02:46:31.000Z,
            "dateUpdated": 2017-05-12T02:46:31.000Z,
            "direction": "outbound-api",
            "from": "+1231231234",
            "sid": "somesid5",
            "status": "delivered",
            "to": "+1231231235",
          },
          Message {
            "body": "foo",
            "dateCreated": 2017-05-12T02:46:30.000Z,
            "dateSent": 2017-05-12T02:46:31.000Z,
            "dateUpdated": 2017-05-12T02:46:31.000Z,
            "direction": "inbound",
            "from": "+1231231235",
            "sid": "somesid0",
            "status": "received",
            "to": "+1231231234",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      onViewableItemsChanged={undefined}
      renderItem={[Function]}
      renderScrollComponent={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "marginBottom": 30,
        }
      }
      updateCellsBatchingPeriod={50}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
        >
          <View
            onLongPress={undefined}
            onPress={undefined}
            style={
              Object {
                "margin": 10,
                "marginRight": 50,
              }
            }
            underlayColor="white"
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "marginLeft": 10,
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "center",
                  }
                }
              >
                <View
                  style={undefined}
                >
                  <View>
                    <View
                      style={
                        Object {
                          "alignSelf": "flex-start",
                          "backgroundColor": "#d05ce3",
                          "borderRadius": 10,
                          "padding": 10,
                        }
                      }
                    >
                      <Text
                        accessible={true}
                        allowFontScaling={true}
                        disabled={false}
                        ellipsizeMode="tail"
                        style={
                          Object {
                            "color": "#ffffff",
                          }
                        }
                      >
                        bar
                      </Text>
                    </View>
                  </View>
                </View>
                <View
                  style={undefined}
                >
                  <View />
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
        >
          <View
            onLongPress={undefined}
            onPress={undefined}
            style={
              Object {
                "margin": 10,
                "marginLeft": 50,
                "marginRight": 30,
              }
            }
            underlayColor="white"
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "marginLeft": 10,
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "center",
                  }
                }
              >
                <View
                  style={undefined}
                >
                  <View>
                    <View
                      style={
                        Object {
                          "alignSelf": "flex-end",
                          "backgroundColor": "#fff",
                          "borderRadius": 10,
                          "padding": 10,
                        }
                      }
                    >
                      <Text
                        accessible={true}
                        allowFontScaling={true}
                        disabled={false}
                        ellipsizeMode="tail"
                        style={
                          Object {
                            "textAlign": "right",
                          }
                        }
                      >
                        foo
                      </Text>
                    </View>
                  </View>
                </View>
                <View
                  style={undefined}
                >
                  <View />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
